build.gradle.kts

```diff
@@ -47,4 +60,76 @@ val integrationTestImplementation: Configuration by configurations.getting {
 }
 val integrationTestRuntimeOnly: Configuration by configurations.getting {
     extendsFrom(configurations.testRuntimeOnly.get())
-}
\ No newline at end of file
+}
+val containerName = findProperty("containerName") ?: "hivemq-k8s-test"
+val containerTag = findProperty("containerTag") ?: "snapshot"
+
+val producerK8sDockerImage: Configuration by configurations.creating {
+    isCanBeConsumed = false
+    isCanBeResolved = true
+    attributes {
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named("k8s-docker-image"))
+    }
+    extendsFrom(hivemq)
+}
+val producerDnsInitWaitDockerImage: Configuration by configurations.creating {
+    isCanBeConsumed = false
+    isCanBeResolved = true
+    attributes {
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named("dns-init-wait-docker-image"))
+    }
+    extendsFrom(operator)
+}
+val producerOperatorDockerImage: Configuration by configurations.creating {
+    isCanBeConsumed = false
+    isCanBeResolved = true
+    attributes {
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named("operator-docker-image"))
+    }
+    extendsFrom(operator)
+}
+
+
+val createRootlessK8sImageContext by tasks.registering(Sync::class) {
+    group = "container"
+    description = "Prepare hivemq rootless k8s image context"
+    into(layout.buildDirectory.dir("container/context"))
+    from("container")
+}
+
+val buildRootlessK8sImage by tasks.registering(Exec::class) {
+    group = "container"
+    description = "Build hivemq rootless k8s image"
+    inputs.property("dockerImageName", containerName)
+    inputs.dir(createRootlessK8sImageContext.map { it.destinationDir })
+    workingDir(createRootlessK8sImageContext.map { it.destinationDir })
+    commandLine("docker", "build", "-f", "broker.dockerfile", "-t", "${containerName}-rootless:${containerTag}", ".")
+}
+
+val saveRootlessK8sImage by tasks.registering(Exec::class) {
+    group = "container"
+    description = "Save hivemq rootless k8s image"
+    dependsOn(buildRootlessK8sImage)
+    workingDir(layout.buildDirectory.dir("containers"))
+    commandLine("docker", "save", "-o", "${containerName}-rootless.tar", "${containerName}-rootless:${containerTag}")
+}
+val buildContainersFiles by tasks.registering(Copy::class) {
+    group = "container"
+    from(producerK8sDockerImage)
+    from(producerDnsInitWaitDockerImage)
+    dependsOn(gradle.includedBuild("hivemq-operator").task(":jibBuildTar"))
+    from(
+        producerK8sDockerImage.singleFile,
+        producerDnsInitWaitDockerImage.singleFile,
+        producerOperatorDockerImage.singleFile
+    )
+    into(layout.buildDirectory.dir("containers"))
+}
+
+buildContainersFiles {
+    finalizedBy(saveRootlessK8sImage)
```

Comment:
I still don't get why the saveRootlessK8sImage must be run after the other tasks.
Maybe a problem is that you have the working directory in the same folder (`workingDir(layout.buildDirectory.dir("containers"))`) but this could be changed.

Reply:
The Dockerfile https://github.com/hivemq/helm-charts/blob/4b4510d41b5eff88b34c50bc5deb53c71e1646c7/container/broker.dockerfile on line 12 uses the image, this dependency will be fix with the new strategy as we expect to have only one image that support both approaches, or we release an image for this

Reply:
Then the buildRootlessK8sImage task should actually depend on the k8s image.
See the above comment, then this can be replaced with a normal dependsOn.